File: src/attestations/signer_sigstore.zig
// SPDX-License-Identifier: MIT OR Apache-2.0
//
// Not yet implemented. Will wrap Fulcio OIDC + Rekor. Tracked in
// docs/SPEC-ATTESTATIONS.md §6.2.
//
// The scaffold exists so the signer router in `mod.zig` can dispatch on
// signer kind and get a concrete `error.SigstoreNotImplemented` back
// instead of a compile-time branch.

const std = @import("std");
const Allocator = std.mem.Allocator;

const signer_mod = @import("signer.zig");
const Signer = signer_mod.Signer;

pub const SigstoreSigner = struct {
    pub fn init() SigstoreSigner {
        return .{};
    }

    pub fn asSigner(self: *SigstoreSigner) Signer {
        return .{ .ptr = @ptrCast(self), .vtable = &vtable };
    }

    const vtable: Signer.VTable = .{
        .keyid = keyidImpl,
        .signDsse = signDsseImpl,
    };

    fn keyidImpl(_: *anyopaque, _: Allocator) anyerror![]u8 {
        return error.SigstoreNotImplemented;
    }

    fn signDsseImpl(_: *anyopaque, _: Allocator, _: []const u8) anyerror![]u8 {
        return error.SigstoreNotImplemented;
    }
};

// -----------------------------------------------------------------------------
// Tests
// -----------------------------------------------------------------------------

const testing = std.testing;

test "sigstore signer: both vtable methods return SigstoreNotImplemented" {
    var ss = SigstoreSigner.init();
    const s = ss.asSigner();

    try testing.expectError(error.SigstoreNotImplemented, s.keyid(testing.allocator));
    try testing.expectError(error.SigstoreNotImplemented, s.signDsse(testing.allocator, "pae"));
}

File: rust/crates/mkit-attest/src/signer_sigstore.rs
//! Sigstore signer — scaffold only.
//!
//! Will eventually wrap Fulcio OIDC + Rekor. Tracked in
//! `docs/SPEC-ATTESTATIONS.md` §6.2. The scaffold exists so a signer
//! router can dispatch on signer kind and get a concrete
//! `Error::SigstoreNotImplemented` back instead of a compile-time
//! branch.

use crate::Error;
use crate::signer::Signer;

#[derive(Debug, Default)]
pub struct SigstoreSigner;

impl SigstoreSigner {
    #[must_use]
    pub fn new() -> Self {
        Self
    }
}

impl Signer for SigstoreSigner {
    fn keyid(&self) -> Result<String, Error> {
        Err(Error::SigstoreNotImplemented)
    }
    fn sign(&mut self, _pae: &[u8]) -> Result<Vec<u8>, Error> {
        Err(Error::SigstoreNotImplemented)
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn both_methods_return_not_implemented() {
        let mut s = SigstoreSigner::new();
        assert!(matches!(s.keyid(), Err(Error::SigstoreNotImplemented)));
        assert!(matches!(s.sign(b"pae"), Err(Error::SigstoreNotImplemented)));
    }
}
